Finished task 3

1337bugs opened the pull request:
Updated get_letter_grades, get_student_average according to the task, introduced a new method update_grades.
Rewritten old methods to ensure they will work correctly if we add more entries to the table.
Error handling should work okay.
NOTE: get_student_average might be improved, is to figure out later.

1337bugs committed:
task3 solution

subtask1 - done
subtask2 - in process

1337bugs committed:
Delete task3_solution.py

1337bugs committed:
Merge pull request #1 from pinlast/master

merge

1337bugs committed:
task3 solution

subtask1 - done
subtask2 - in process, figure out how to process the exception correctly
subtask3 - not done

1337bugs committed:
test commit task3

1337bugs committed:
Task 3 in progress

subtask 2 almost done

1337bugs committed:
Task 3 update

updated `get_letter_grades` and `get_student_average` according to the task and got it working

1337bugs committed:
Finished task 3

Updated get_letter_grades, get_student_average according to task, introduced new method update_grades.
Rewritten old methods to ensure they will work correctly if we add more entries to the table.
Error handling should work okay.
NOTE: get_student_average might be improved, is to figure out later.

1337bugs committed:
Update task3_solution.py

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,150 @@
+                break
+        return found
+
+    def get_letter_grades(self, *students, throw_exception=False):
```
args can be called everything you want but they must always be args or in case of kwargs -- kwargs, so everyone always knows what this is about

1337bugs replied:
python docs and tutorials online misled me a bit about naming *args, will consider this a must

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,150 @@
+    def get_letter_grades(self, *students, throw_exception=False):
+        for name in students:
+            # if name is present in dictionaries, do stuff
+            if self.inDict(name):
```
You can use .get() here, it's much simpler and there is no need for another method

pinlast replied:
smth like if  `self.student_data.get(name): do stuff`

pinlast replied:
and all method should be in snake case
Camel case only for classes

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,150 @@
+            # if name is present in dictionaries, do stuff
+            if self.inDict(name):
+                for dict in self.student_data:
+                    if dict.get("name") == name:
```
This is really overcomplicated, try using one `.get()`

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,150 @@
+                for dict in self.student_data:
+                    if dict.get("name") == name:
+                        print("Grades of " + name + ": ")
+                        temp = dict.copy()
```
There is no need for copy and pop, easier way would be just add `if` in loop

1337bugs replied:
thought about this, but didn't figure out how to skip first key-value pair using `if`, this seemed the okayish way

pinlast replied:
You can't skip first key in dict, because it's not sorted. Any key can be in any position. You should look for "name" key exactly

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,150 @@
+                        for values in temp.values():
+                            list.extend(values)
+                        print("Average grade of", name, ":",
+                              "{0:5.2f}".format(self.avg_grade(list)))
```
1) avoid using format, try f-string
2) if you always use {.2f} for avg_grade, try moving it to method itself

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,150 @@
+                    print("No entry", name, "in table")
+
+    # to see the state of data
+    def test(self):
```
test isn't good name for method, as name suggests that method should test something, but it just prints values

1337bugs replied:
this was a throwaway method initially, have no idea why I left it in PR, even uncommented

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,150 @@
+    # to see the state of data
+    def test(self):
+        for data in self.student_data:
+            print([value for key, value in data.items()])
```
You can just use `data.values()`

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,150 @@
+        if self.inDict(student_name):
+            for dict in self.student_data:
+                # seems like im breaking DRY next line
+                if dict.get("name") == student_name:
```
same thing with really complicated validation if item in dict

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,150 @@
+                # seems like im breaking DRY next line
+                if dict.get("name") == student_name:
+                    for new_work, new_grades in kwargs.items():
+                        for work, grades in dict.items():
```
You can just do `dict[new_work].extend(new_grades)`

pinlast replied:
And overall main reason to use dicts is that you can just grab stuff using key with O(1), looping over them usually can be avoided with `.get()` or some other built in method

1337bugs replied:
yes, good point. I've read about dict features and learnt that you don't always need to iterate, but I need to get a hand in using them. is it a good practice just to use as many expressions with `.get()` as you need? 
also I don't realize why `.get()` can't accept a variable (e.g. string or another hashable) as an argument, but rather takes a string directly - `.get("homework")` works fine, `.get(new_work)` doesn't. faulty code or a misconception?

pinlast reviewed (commented):
Overall you overcomplicated it too much, I added lot of comments, but it's okay, those are pretty logical errors, you'll avoid them with more practice.

1337bugs committed:
Task 3 update

almost finished fixing stuff, now code is simpler, though there is some logical errors to be fixed

1337bugs commented:
the code is simpler now, though I cant seem to find good ways to resolve logical errors. see the output for that.

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,124 @@
+            res += " "
+        return res.strip()
+
+    # calc overall average grade
```
To explain what functions does it's better to use docstrings
https://www.geeksforgeeks.org/python-docstrings/

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,124 @@
+
+    def update_grade(self, student_name, throw_exception=False, **kwargs):
+        for dict in self.student_data:
+            if student_name in dict.get("name"):
```
dict.get("name") returns None if "name" was not found, and you are looking for student_name in None
You can simply make dict.get("name")

pinlast replied:
And dict is built in function name, it's better to change it to something more specific to that case

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,124 @@
+                    dict[new_work].extend(new_grades)
+            else:
+                if throw_exception:
+                    print("No entry", student_name, "in table")
```
Try using f-string for cases when you need put variable in string

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,124 @@
+
+    def get_student_average(self, *args, throw_exception=False):
+        for dict in self.student_data:
+            if dict.get("name", 0) in args:
```
0 doesn't seem to fit here, you are looking for strings and getting 0 to search for
You can either leave it as default None, or change it to string

1337bugs replied:
This `if` is meant to evaluate to 0 if name is not found, therefore executing `else` statement with `throw_exception`, and that's where I'm experiencing troubles.

pinlast replied:
Yeah, but you can just leave `None` or `''` and you'll get same result but it would make much more sense then `0` in that case

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,124 @@
+
+    # to see the state of data
+    def show_data(self):
+        for data in self.student_data:
```
It's not really important in that case but try using more meaningful names for variables

pinlast reviewed (changes requested):
Added some small comments, just to make it better overall
But it's actually pretty good already

1337bugs committed:
Update task3_solution.py

minor improvements

pinlast commented on task3/task3_solution.py:
```diff
@@ -0,0 +1,130 @@
+
+    def get_student_average(self, *args, throw_exception=False):
+        for entry in self.student_data:
+            if entry.get("name", None) in args:
```
There is no need to put `None` as second `arg`, it's default to `None`
